### PR TITLE
Fix exceptions on empty tuple errors

### LINKED ISSUE #17266: Debug Failure when writing Promise<[]>
captura de tela de 2017-07-18 12-13-27
captura de tela de 2017-07-18 12-16-35

**TypeScript Version:**  2.4.1

Expected behavior:
There should be no error in this.

Actual behavior:
When returning Promise<[ ]> without the type, an error occurs in the tsc(Debug failure).

**code**
```javascript

export async function get(): Promise<[]> {
  let emails = [];
  return emails;
}

```
error message: 

```shell
yarn tsc
$ tsc
/home/jean/abrafarma/node_modules/typescript/lib/tsc.js:55776
                throw e;
                ^

Error: Debug Failure.
    at typeToString (/home/jean/abrafarma/node_modules/typescript/lib/tsc.js:23526:22)
    at reportRelationError (/home/jean/abrafarma/node_modules/typescript/lib/tsc.js:28855:34)
    at isRelatedTo (/home/jean/abrafarma/node_modules/typescript/lib/tsc.js:28972:21)
    at checkTypeRelatedTo (/home/jean/abrafarma/node_modules/typescript/lib/tsc.js:28838:26)
    at checkTypeAssignableTo (/home/jean/abrafarma/node_modules/typescript/lib/tsc.js:28578:20)
    at checkReturnStatement (/home/jean/abrafarma/node_modules/typescript/lib/tsc.js:37576:33)
    at checkSourceElement (/home/jean/abrafarma/node_modules/typescript/lib/tsc.js:38751:28)
    at Object.forEach (/home/jean/abrafarma/node_modules/typescript/lib/tsc.js:298:30)
    at checkBlock (/home/jean/abrafarma/node_modules/typescript/lib/tsc.js:36918:16)
    at checkSourceElement (/home/jean/abrafarma/node_modules/typescript/lib/tsc.js:38730:28)
error Command failed with exit code 1.

```
TsConfig:
```json
{
  "compilerOptions": {
    "target": "es2017",
    "module": "commonjs",
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "removeComments": false,
    "noImplicitAny": false,
    "noEmitOnError": true,
    "importHelpers": true,
    "alwaysStrict": true,
    "inlineSourceMap": true,
    "inlineSources": true,
    "rootDir": "server",
    "outDir": "dist_server",
    "noUnusedLocals": true
  },
  "include": [
    "server/**/*.ts"
  ],
  "compileOnSave": false,
  "exclude": [
    "node_modules",
    "bower_components"
  ]
}


```

### LINKED ISSUE #17303: Crashes on invalid 2D-array declarations
Took me ages to find out that when by mistake I declared a 3D-array as `any[[]]` rather than `any[][]`, TypeScript compiler (any version) just crashes with:
```
Error: Debug Failure.
```

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #17266
Fixes #17303

Our type to string function expected the node builder to always yield a type when errors were ignored, when for empty tuples it still wasn't returning nodes for an empty tuple type. Now it does.

### CODE REVIEW

**Comment by weswigham** [comment]
I don't know how to restart the `.NET` CI builds like I can travis; but the build is functional again.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -2623,8 +2623,9 @@ namespace ts {
                         }
                         if (!context.encounteredError && !(context.flags & NodeBuilderFlags.AllowEmptyTuple)) {
```
Comment (sandersn) [inline src/compiler/checker.ts]: Invert the predicate here to make it more readable and put `return undefined` at the end.

**Comment by sandersn** [comment]
One NIT to fix.

Also, I think that Jenkins might respond to commands in comments, if you know who to @ and you have permissions and you know the command.
@RyanCavanaugh probably knows the answer to these but it's probably not worth it.

**Comment by pmunin** [comment]
Thanks for the fix! I see it was merged to master, but I still experience this on latest version of Typescript (2.4.2) that I installed via npm. When can we expect it to be updated in NPM?

**Pull request merged**